include/seqan3/range/view/single_pass_input.hpp

```diff
@@ -0,0 +1,434 @@
+     */
+    //!\brief Default constructor.
+    single_pass_input_sentinel() = default;
+    //!\brief Copy constructor.
+    constexpr single_pass_input_sentinel(single_pass_input_sentinel const &) = default;
+    //!\brief Move constructor.
+    constexpr single_pass_input_sentinel(single_pass_input_sentinel &&) = default;
+    //!\brief Copy assignment.
+    constexpr single_pass_input_sentinel & operator=(single_pass_input_sentinel const &) = default;
+    //!\brief Move assignment.
+    constexpr single_pass_input_sentinel & operator=(single_pass_input_sentinel &&) = default;
+    //!\brief Destructor.
+    ~single_pass_input_sentinel() = default;
+
+    //!\brief Construction from underlying sentinel.
+    single_pass_input_sentinel(sentinel_t _s) : sentinel{_s}
+    {}
+    //!\}
+
+    /*!\name Comparison
+     * \{
+     */
+    //!\brief Comparison with iterator type.
+    template <typename view_t>
+    bool operator==(single_pass_input_iterator<view_t> const & rhs) const noexcept
+    {
+        return sentinel == rhs.cached();
+    }
+
+    //!\brief Comparison with iterator type.
+    template <typename view_t>
+    bool operator!=(single_pass_input_iterator<view_t> const & rhs) const noexcept
+    {
+        return !(*this == rhs);
+    }
+    //!/}
+};
+} // seqan3::detail
+
+//-----------------------------------------------------------------------------
+// Implementation of single pass input view.
+//-----------------------------------------------------------------------------
+
+namespace seqan3::detail
+{
+
+/*!\brief Adds single_pass_input behavior to the underlying range.
+ * \tparam urng_t The underlying range type.
+ * \implements input_range_concept
+ * \ingroup view
+ */
+template <typename urng_t>
+//\cond
+    requires input_range_concept<std::remove_reference_t<urng_t>>
+//\endcond
+class single_pass_input_view : public detail::view_base
+{
+private:
+
+    //!\brief Friend declaration for seqan3::detail::single_pass_input_iterator.
+    template <typename view_t>
+    friend class single_pass_input_iterator;
+
+    //!\brief The pure range type without any reference type.
+    using pure_range_t       = std::remove_reference_t<urng_t>;
+    //!\brief The iterator type for the underlying range.
+    using urng_iterator_type = ranges::v3::iterator_t<pure_range_t>;
+    //!\brief The sentinel type for the underlying range.
+    using sentinel           = single_pass_input_sentinel<ranges::sentinel_t<pure_range_t>>;
+
+    //!\brief The internal state.
+    struct state
+    {
+        //!\brief The underlying range.
+        urng_t             urng;
+        //!\brief The cached iterator of the underlying range.
+        urng_iterator_type cached_urng_iter{};
+    };
+
+    //!\brief Shared pointer of the data model.
+    std::shared_ptr<state> state_ptr{};
+
+public:
+    /*!\name Member types
+     * \{
+     */
+    //!\brief Reference type.
+    using reference         = ranges::range_reference_t<pure_range_t>;
```

Comment:
change to:
```cpp
using reference = std::conditional_t<std::is_const_v<urng_t>,  const_reference_t<uncvref_t<urng_t>>, reference_t<uncvref_t<urng_t>>>;
```

depends on #159

Reply:
Well, single pass input cannot return a ref, because you can possibly not write into the input object, e.g. a stream.
I however, also added const to the reference type to ensure, that also a vector becomes unwritable through this view.